Added needed functionality for SNICKER

AdamISZ opened the pull request:
(Updated)
See next comment for info on how to test this.

This builds on the earlier SNICKER code (see [spec](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/768/commits/b545986cb5df1798dc71b91cdaa2a917638bbd5c)) that was added back in #536 (it was only added at this point because only after `python-bitcointx` did we have PSBT functionality provided).

The branch is called `snicker-mvp` (as in minimum viable product) with reason; it provides only a simplified, manual workflow. It is intended to allow testing that would hopefully allow us to add this functionality, initially non-default, to our wallets and bots.

What is added:
* Subdirectory `scripts/snicker` for the following user tools:
* `snicker-seed-tx.py` - create a fake SNICKER transaction in your own wallet
* `snicker-finder.py` - scan recent blocks for Joinmarket or SNICKER candidate transactions
* `create-encrypted-proposal.py` - takes transactions found from the above and makes proposals, uploading them to a server
* `snicker-server.py` - implements a simple server serving over *.onion, with a sqlite database to store proposals, and defends against spam only mildly with a proof of work requirement (see below)
* `receive-snicker.py` - polls above server to read new proposals (or just accept manually provided proposals as argument), parse them and broadcasts completed SNICKER coinjoins when found, storing the new keys as imports (see details on wallet handling below).
*  `snicker-recovery.py` - can be used to recover a wallet from seedphrase which contains SNICKER utxos, though it needs (possibly multiple) rescanblockchain calls (and informs the user how to do this, including blockheights).
 * New config section `[SNICKER]` as follows (see `jmclient.configure` for comments/explanation of fields):  `enabled = false`, `lowest_net_gain = 0`, `servers = cn5lfwvrswicuxn3gjsxoved6l2gu5hdvwy5l3ev7kg6j7lbji2k7hqd.onion,`, `polling_interval_minutes = 60`
 * New AMP commands in `jmbase.commands` for SNICKER, and correspondingly:
 * New `SNICKERClientProtocol` and `SNICKERDaemonServerProtocol` protocols to support sending requests and responses back and forth between `jmclient` and `jmdaemon` (to preserve the principle of separating external network interaction from the business logic (and bitcoin logic) of the application (more on this below).
 * New base protocol type `HTTPPassThrough` as parent class of the above `SNICKERClientProtocol` as a useful refactoring: we will want basically this exact same functionality for Payjoin, in order to refactor it also to have external network interaction only in daemon (I plan to do that after this is merged).
 * Closely related to the above, some Tor network code (using `txtorcon` to start ephemeral onion service, and using twisted `Agent` objects to instantiate client connections to both onion services and TLS endpoints) is refactored into `jmbase.twisted_utils` so it can be reused.
 * `yieldgenerator` will start an instance of the `SNICKERReceiver` via the `SNICKERClientProtocol` **if** the config says SNICKER is enabled (it will not, by default), thus a Maker will be able to be a SNICKER receiver.
 * A little bit of extra functionality in `jmbitcoin.secp256k1_transaction` : witnesses and in particular pubkeys can be extracted from signed transactions (this of course is needed for SNICKER v1 but was added to `jmbitcoin` as it could be useful at other times).
 * A number of other changes were made in the `SNICKERWalletMixin` and the `SNICKERReceiver` objects in jmclient, basically completing the desired functionality, e.g. ensuring native segwit was fully supported and output keys were correctly imported.

AdamISZ commented:
(Placeholder: will add detailed testing instructions here)

AdamISZ commented:
A rough sketch of what I think will be needed before moving this from [WIP] to ready-to-merge:

* Update `jmclient/test_snicker.py` for the code changes
* Implement a logs/SNICKER.log at minimum, so a record of the SNICKER coinjoins a receiver does are always available.
* Although a fully viable server might be out of scope here (I want someone else to write it, really!), at least update the server to keep its DB size limited (ideally add some crude anti-DOS as well).
* Fix the bug where `create-snicker-proposal.py` hangs after completing all POST requests successfully.
* Fix the bug that the config file entry for `servers` must use the `http://` prefix (if *.onion not https).
* Add option to always prefer spending imported coins (as SNICKER outputs will be), to minimize impact of non-recoverability from seed
* Add a recovery script based on the principle described roughly in the spec: if historical transactions are retrieved, SNICKER outputs can be detected and the keys rederived. This is a non-trivial script however.

(Also a TODO related to this PR, but not part of it: once the `HTTPPassthrough` protocol is merged, we can remove the networking code from `jmclient.payjoin` by following the same procedure: create a `PayjoinClientProtocol` and a `PayjoinServerProtocol` to basically pass through the sender's requests. Receiver will be handled similarly, though not yet sure of details.)

AdamISZ commented:
This is much more functional now, enough so that it is worth testing. Read https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/f6ca59dd58a80decccae5db07896c12cea17f413/docs/TESTING-SNICKER.md for details and instructions on currently how to test.

AdamISZ commented:
See https://github.com/meejah/txtorcon/issues/347 in relation to https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/768/commits/78b05e1c4b0c461991487f2beb124f3495ad7149 - although this will not be relevant for testing, or for non-server runners.

AdamISZ commented:
Feels like this "fell at the final hurdle" - in as much as, it seems very difficult to effect the final bullet point in [this](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/768#issuecomment-746888775) comment, i.e. the recovery script.

Everything apart from that recovery seems to be fully functional (and I'm running the server at the configured onion address, if anyone wants to try it).

I will note down how far I've got with this recovery script, and I'm leaving it up in the air whether it can be finished:

The algorithm, as roughly noted in the original gist, is something like this:

* Run wallet sync in recovery mode. Since it's recovery from scratch, it does need (hopefully only one) rescan in Core.
* After rescan finishes and recovery mode sync completes, you should have all transactions concerning addresses in the HD tree available via RPC.
* Find the transactions which spend from the above, but don't receive. This is one very tricky point: I have found that the RPC `listtransactions` will *not* show send-category transactions if you specify a particular wallet label (joinmarket-XXXXX); but if you specify no label with "*", it does show send-category transactions. If this is 100% correct then, OK; we can still always find those transactions. I have been able to do it in some rudimentary tests.
* Next, you need to filter that list of txs for `is_snicker_tx()`, which is easy enough. Then check if both of the equal sized outs are not in our HD tree, and if so check against the inputs of the tx which are not ours, and the inputs of the *ancestor* txs of the inputs which *are* ours, and try ECDH on each of these pubkey pairs, to check if a tweak can generate one of the output addresses as according to the spec. Although this sounds complicated, it's fairly trivial to code.
* Assuming all the above works fine, here's where I think we hit the actual stumbling block: the SNICKER transactions identified above (assuming you do succeed in finding a matching tweaked output), will produce a new txo that we owned, but without that having been pre-imported before rescan, we can't find the transaction which spends that txo. This would mean there isn't a trivial way to "search forwards" - imagine a scenario of a whole chain of SNICKER transactions, each one spending the tweaked output (i.e. the receiver coinjoin output) of the previous one. What is needed is some kind of "get_tx_spending_this_output" rpc call or similar. Without fully scanning blocks, how can this be done?

harding commented:
@AdamISZ 

> This would mean there isn't a trivial way to "search forwards" - imagine a scenario of a whole chain of SNICKER transactions, each one spending the tweaked output (i.e. the receiver coinjoin output) of the previous one. What is needed is some kind of "get_tx_spending_this_output" rpc call or similar. Without fully scanning blocks, how can this be done?

The fastest way I can think to accomplish this without using just Bitcoin Core would be to require creating the BIP158 compact block filter index and then to use that for your forward scan.  For example, you know you received coins to `address1` and `address2`, the earliest of these was received at block `x`.  So you compare {`address1`, `address2`} to the filter for block `x+1`; if there's no match, you compare again to block `x+2`.  If a filter matches, you scan that whole block and, if it's a true positive, extract the transaction and replace the used address (e.g. `address1`) with the new `address3`, then scan from the next block.  If it's a false positive, you continue to the next block like normal.

The existing `getblockfilter` RPC returns the per-block filters and I think writing a GCS parser should be pretty easy (and GCS decoding and comparison should be very fast).  However, I did a quick test and the `getblockfilter` RPC is quite slow using a simple shell for loop, so maybe that's a problem.  It could maybe be solved with RPC request batching or patching the RPC to enable its own ability to batch multiple filters together in one response.

AdamISZ commented:
Hi David, thanks for your thoughts. First thing I'm wondering is what you meant by this:

> The fastest way I can think to accomplish this without using just Bitcoin Core

why "without using just Bitcoin Core"? The scenario I'm describing is restoring a Joinmarket wallet from a seedphrase, using Bitcoin Core (with watch only addresses), i.e. how Joinmarket normally works.

harding commented:
Whoops, that was a typo; I edited my post and deleted "without".

AdamISZ commented:
So I spent a bit of time reading up on the BIP157/158 mechanism and I can see that @harding 's suggestion here makes a lot of sense, at least as I understand it. The recovery script would do it's "forward-scan" on HD addresses as normal, followed by looking at each SNICKER-receipt imported scriptpubkey, for which the utxo is spent but we don't know where it is spent. As explained in the 5th bullet point above, we have the issue of how to find those spends, in general.
So we would indeed be able to use the basic filters against those scriptpubkeys for each block, and download the block that spends it, find it, and then iterate forwards as/if needed, and as per @harding's suggestion. Even if it is a bit slow, it will be *substantially* faster than parsing all blocks, but I guess the other option is just to require txindex for anyone who needs to do this (rare) recovery case?

A couple of practical points crop up if we go down this route:
* do we need to implement bitcoin p2p messaging? presumably we do; is there any python library existent that could do that (wouldn't be too hard but don't want to reinvent wheels)?
* if we embark on including this feature, does it make sense to include it for a general light-client mode of Joinmarket, or is that (a) too much work or (b) too suboptimal in other ways, e.g. privacy or functionality provided by full/pruned node?
* We'll obviously also need the GCS matching set to filter code, as per BIP158; again, is there an extant Python library or module handling that?

AdamISZ commented:
So, to follow this up:
I spent a lot of time reading up on GCS and BIP158 filters, I think it's a very promising technology, which could indeed make recovery from seedphrase at least *somewhat* more performant in this edge case mentioned, but I've decided not to pursue it right now as there seems to be a lot of Python code required.

For now, in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/768/commits/2e821f716588bfe9b30e2e6fc700f0a73666c220 I have added a script which iteratively scans-and-imports addresses for any SNICKER outputs created at "one step removed" from the existing wallet, then instructs the user to `rescanblockchain` from height N, where N is the earliest block containing one of these SNICKER transactions, followed by repeatedly running this recovery script as many times needed. Except in rare cases this should succeed in finding all coins in a few tries. A reminder that this is the exceptional case of a person losing everything (including the Core wallet instance) but the seedphrase *and* having used SNICKER *and* having used it quite a lot. Even so, it should still work, just be somewhat slow.

I'll be putting some final touches to this PR shortly. Every tool needed for SNICKER I believe works here, now.

AdamISZ commented:
Rebased on master as of https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/768/commits/3870930dca3388ee664e88e8205bb9068cca2f52

AdamISZ commented:
I believe this is now functional and is in a very testable state (see the detailed instructions in the `docs/TESTING-SNICKER.md` doc). So I've removed the [WIP] from the title (and also the 'basic form' - that may technically be true, but there is a *lot* of functionality here).

AdamISZ commented:
OK after a brief review of the state of the repo, I'm inclined to rebase, touch this up a bit then merge it. The main concept would be: it's off by default (obviously, as mentioned above), and the docs for it specify to use it for signet(testnet) only, for now. Whilst this would be interesting for itself, and I'd hope people try it out like that, there's also the fact that it allows us to finally address #748 - the daemon now controlling the tor HS code, as described there. There's also some extra features generally useful in here, including a scanner for joinmarket coinjoins (slow, but of niche interest).

As previously any code review, comments or tests would be very helpful.

AdamISZ commented:
Rebased on master as of https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/ff948bcaf9b3ceab88edf1d5c32bb83bb5ac8a03

AdamISZ commented:
Seems to work fine on signet, this is an example from following the "single user test workflow" in the doc: 13ba85c886d09eaddea75c9cc706dfb9084759f93903423087c31b86056effdc (you can see that it was funded and triggered by a fake snicker created with snicker-seed-tx.py).

MaxHillebrand commented:
Thanks @AdamISZ for picking this up again, promising work, thanks! :black_heart: 

Just for the record, here's a [SigNet block explorer](https://signet.btc21.org/tx/13ba85c886d09eaddea75c9cc706dfb9084759f93903423087c31b86056effdc), to save y'all some time in finding one ;)

AdamISZ commented:
Yeah sorry sloppy of me, that block explorer is linked in my gist about Joinmarket with signet, which is linked in the release notes, but buried under two layers of link is not ideal.

AdamISZ commented:
https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/768/commits/0e8e33817c6501dd1083debeb7bbbfcfc5ce8928 is a complete rewrite of the SNICKER.md documentation; it should hopefully be a lot more readable/sensible/approachable for the general reader, now.

PulpCattel commented on docs/SNICKER.md:
````diff
@@ -0,0 +1,322 @@
+```
+... to a `joinmarket.cfg` that you add inside `scripts/snicker`, by copying it from `scripts/` or wherever you keep your testing `joinmarket.cfg` file. (This manual annoyance is part of testing, it won't be needed in mainnet usage of course ... alternatively just put a signet/regtest custom `joinmarket.cfg` in your `.joinmarket` folder, but this is hardly less annoying).
+
+`servers=` there is a comma separated list.
````
```suggestion
`servers=` accepts a comma separated list.
```
Possible suggestion, also maybe `requires` could work.
This whole line is redundant with the dedicated paragraph about SNICKER settings, but I guess in a long doc file a repetition doesn't hurt.

PulpCattel commented on docs/SNICKER.md:
```diff
@@ -0,0 +1,322 @@
+
+`servers=` there is a comma separated list.
+
+You're now ready to do the two steps: (a) create a proposal and upload it, (b) download proposals (as the receiver identity/wallet) and complete coinjoins. It could be different people doing (a) and (b) of course but here we're assuming one tester doing everything (see two wallets above).
```
Here a markdown list could be nice and clear. 

* a) Create a proposal and upload it.
* b) Download proposals (as the receiver identity/wallet) and complete coinjoins.

It could be different people doing (a) and (b) of course but here we're assuming one tester doing everything (see two wallets above).

AdamISZ replied:
Not needed

PulpCattel commented on docs/SNICKER.md:
````diff
@@ -0,0 +1,322 @@
+```
+python create-encrypted-proposal.py --datadir=. proposerwallet.jmdat "0200000000010138e8a90b3df7740b9d5f5ae9af2cf6769f314d290b2e12bf25bfa4aae2c0cbe20000000000feffffff0280ba8c010000000016001471b09afbac6204627225c10f3a8d4a0749364fdb6d7c36220000000016001447ae59f32c504cbb56e18b77f7842fb58b55025b02473044022075354351ad4c619ba662f9abd25e8ee434f8381795001606a29fa959d36aeb7f022018f8bf1ec0407dad586baeb7e4d ...
+```
+Obviously see the `--help` for details, but in this example we chose input index 0 to source the pubkey for the encryption, output index 1 for the coin we want the receiver to spend with us in the coinjoin, 100 sats as the amount we will bump their output by as an incentive to do the coinjoin (you *can* make this number negative, to receive, instead - the proposer is paying the tx fee otherwise,  ...
````
Here too a list could come in handy:

---

Obviously see the `--help` for details, but in this example we chose:

* Input index 0 to source the pubkey for the encryption.
* Output index 1 for the coin we want the receiver to spend with us in the coinjoin.
* 100 sats as the amount we will bump their output by as an incentive to do the coinjoin (you *can* make this number negative, to receive, instead - the proposer is paying the tx fee otherwise, note).
* Mixdepth 1 as the mixdepth from which we source *our* coins for the coinjoin. Make sure of course that mixdepth 1 has at least a little bit more bitcoin than the size of that output at index 1 aforementioned.
* The `-n` option can be used to output the proposal to stdout (it is base64+hex so copy-pasteable).

---

Also, `you *can* make this number negative, to receive, instead - the proposer is paying the tx fee otherwise, note` could be rephrased as to me seems a bit confusing.

AdamISZ replied:
Agree on this one

PulpCattel commented on docs/SNICKER.md:
````diff
@@ -0,0 +1,322 @@
+```
+Obviously see the `--help` for details, but in this example we chose input index 0 to source the pubkey for the encryption, output index 1 for the coin we want the receiver to spend with us in the coinjoin, 100 sats as the amount we will bump their output by as an incentive to do the coinjoin (you *can* make this number negative, to receive, instead - the proposer is paying the tx fee otherwise,  ...
+
+If you choose not to use -n, but instead use a proposals server as above, then assuming it connects the server OK, you will see:
````
```suggestion
If you choose not to use -n, but instead use a proposals server as above, then assuming it connects to the server OK, you will see:
```

PulpCattel reviewed (commented):
Just proofreading the SNICKER docs.

There are several paragraphs that, in raw text, are super long, could be nice to split them over multiples lines.

AdamISZ commented on docs/SNICKER.md:
````diff
@@ -0,0 +1,322 @@
+The unspent indices are: 0 1 2
+```
+
+Copy that fully signed transaction hex, and note the unspent outputs. You're going to assume (in this case correctly of course) that all of the inputs are valid options for the SNICKER public key (why? because it was a *seed* transaction, it was a fake SNICKER, and so all the inputs belonged to the same wallet).
````
not needed

AdamISZ commented on docs/SNICKER.md:
```diff
@@ -0,0 +1,322 @@
+<a name="pow" />
+### Proof of work for mild anti-spam
+
+As implemented here, in fact, the proposer attaches a proof of work in the form of a 10-byte nonce appended to the end of the above string (hex encoded; so in fact: `base64-encoded-ECIES-encrypted-PSBT,hex-encoded-pubkey,hex-encoded-nonce` is what is sent over the wire). This nonce is grinded to get a 32-byte-truncated-hash512 of that string to be less than a target calculated by a request number ...
```
Not needed

AdamISZ commented:
Thanks for those suggestions. To save time I just applied them without comments, except for the ones I disagreed with or thought weren't needed. For the large paragraphs, sorry but I don't think it's worth it, seems nbd.

AdamISZ commented:
https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/768/commits/660e5867390a5df42a54b823c7f557b13e6c6929 corrects a pretty egregious bug; interestingly it wasn't found by the test module because that only checked correct balance changes (and therefore didn't notice address reuse). This is a good example of end to end testing on something like signet being useful, because the block explorer interface made it much easier for me to eyeball every detail of the created transactions (while on regtest I'm mostly stuck with bitcoin-cli and the similar text outputs offered by JM itself).

AdamISZ commented:
Successful test of recovery:

Had a signet wallet that I'd done a couple of JM coinjoins on, then created two fake and two real SNICKER coinjoins (plus of course a couple of deposits from faucet), over a few days. Was able to test recovery in the following way:

* Note the seedphrase of the existing wallet with `python wallet-tool.py walletname showseed`.
* Create/load a new Core wallet with `bitcoin-cli -signet loadwallet testrecovery`.
* Edit `joinmarket.cfg` so `rpc_wallet_file = testrecovery` (as opposed to what I was using for all the above activity, a different wallet called `jm_2`).
* Now we've switched to a new Core wallet which doesn't know about the history, do `python wallet-tool.py recover` and enter the aforementioned seedphrase, and create a new JM wallet `test_snicker_recovery.jmdat` (e.g.).
* At this point you should do a rescan from whenever the original JM wallet was created, or before that, i.e. `bitcoin-cli -rpcwallet=testrecovery rescanblockchain 24000` (note `-rpcwallet=` setting; it must be the newly created and loaded Core wallet).
* Now run the new recovery script: `python snicker/snicker-recovery.py test_snicker_recovery.jmdat --recoversync` (TODO enforce `--recoversync` here as it makes no sense without it).

In the concrete case I mentioned at the start, this succeeded in finding everything after 2 rounds of "run script, then rescan". Here is the output from the script the two times:

```
(jmvenv) waxwing@here~/testjminstall/joinmarket-clientserver/scripts/snicker$ python snicker-recovery.py --recoversync testpostsnicker1.jmdat
User data location: /home/waxwing/.joinmarket/
2021-02-17 16:04:32,359 [DEBUG]  rpc: getblockchaininfo []
Enter passphrase to decrypt wallet: 
2021-02-17 16:04:35,297 [DEBUG]  requesting detailed wallet history
2021-02-17 16:04:35,339 [DEBUG]  rpc: listlabels []
2021-02-17 16:04:35,339 [DEBUG]  rpc: getaddressesbylabel ['joinmarket-wallet-9d5804']
2021-02-17 16:04:35,343 [DEBUG]  got used indices: {0: [3, 6], 1: [0, 2], 2: [0, 0], 3: [0, 0], 4: [0, 0]}
2021-02-17 16:04:35,385 [DEBUG]  rpc: listlabels []
2021-02-17 16:04:35,386 [DEBUG]  rpc: getaddressesbylabel ['joinmarket-wallet-9d5804']
2021-02-17 16:04:35,387 [DEBUG]  Syncing iteration finished, additional step required (gap limit used)
2021-02-17 16:04:35,387 [DEBUG]  rpc: listunspent [0]
2021-02-17 16:04:35,389 [DEBUG]  bitcoind sync_unspent took 0.002066373825073242sec
2021-02-17 16:04:35,390 [DEBUG]  requesting detailed wallet history
2021-02-17 16:04:35,440 [DEBUG]  rpc: listlabels []
2021-02-17 16:04:35,441 [DEBUG]  rpc: getaddressesbylabel ['joinmarket-wallet-9d5804']
2021-02-17 16:04:35,446 [DEBUG]  got used indices: {0: [3, 12], 1: [0, 2], 2: [0, 0], 3: [0, 0], 4: [0, 0]}
2021-02-17 16:04:35,488 [DEBUG]  rpc: listlabels []
2021-02-17 16:04:35,489 [DEBUG]  rpc: getaddressesbylabel ['joinmarket-wallet-9d5804']
2021-02-17 16:04:35,490 [DEBUG]  Syncing iteration finished, additional step required (gap limit used)
2021-02-17 16:04:35,490 [DEBUG]  rpc: listunspent [0]
2021-02-17 16:04:35,492 [DEBUG]  bitcoind sync_unspent took 0.0020096302032470703sec
2021-02-17 16:04:35,493 [DEBUG]  requesting detailed wallet history
2021-02-17 16:04:35,546 [DEBUG]  rpc: listlabels []
2021-02-17 16:04:35,546 [DEBUG]  rpc: getaddressesbylabel ['joinmarket-wallet-9d5804']
2021-02-17 16:04:35,552 [DEBUG]  got used indices: {0: [3, 12], 1: [0, 2], 2: [0, 0], 3: [0, 0], 4: [0, 0]}
2021-02-17 16:04:35,593 [DEBUG]  rpc: listlabels []
2021-02-17 16:04:35,594 [DEBUG]  rpc: getaddressesbylabel ['joinmarket-wallet-9d5804']
2021-02-17 16:04:35,595 [DEBUG]  Wallet successfully synced
2021-02-17 16:04:35,595 [DEBUG]  rpc: listunspent [0]
2021-02-17 16:04:35,597 [DEBUG]  bitcoind sync_unspent took 0.0020742416381835938sec
Found a snicker tx: 0afea91a2f9b0ae4e67ef45ea5f1e3e72a7866070de6394cc5c9343eb6518af8
Found a snicker tx: a54fddd20f2cb2cce051b8e7d6138cdd8154ecb5189e6a692c6d7d1dae2bc301
Found a snicker tx: 13ba85c886d09eaddea75c9cc706dfb9084759f93903423087c31b86056effdc
Found a new SNICKER output belonging to us.
Output address tb1q8znprh8c85za3mpwzn3qf9m0vwqzjkfu4qdncy in the following transaction:
{
    <tx snipped>
}
Importing the address into the joinmarket wallet...
... success.
2021-02-17 16:04:35,641 [DEBUG]  rpc: getblockheader ['0000006f10b5768df19cf9524cd8c89e9fb951dc6d8d9a49915de8e006f2289d']
Found a snicker tx: 762a697e647ca7642317c860508dfd92829a59a180b35cc0c7047c77ad696c15
Found a new SNICKER output belonging to us.
Output address tb1qg7saqx69yalcqshfr8mjndy0gpx2umxrwqs823 in the following transaction:
{
    <tx snipped>
}
Importing the address into the joinmarket wallet...
... success.
2021-02-17 16:04:35,663 [DEBUG]  rpc: getblockheader ['0000014abff83d54692890f394fd3262cf4d4557551b049129cf95ae335e6b6f']
New SNICKER addresses were imported to the Core wallet; do rescanblockchain again, starting from block 25165, before restarting this script.
Script finished, recovery is NOT complete.
```

Notice it tells the user exactly what height to rescan from, which should save significant time. So I did `bitcoin-cli -signet -rpcwallet=testrecovery rescanblockchain 25165`, and then, second run of script:

```
(jmvenv) waxwing@here~/testjminstall/joinmarket-clientserver/scripts/snicker$ python snicker-recovery.py --recoversync testpostsnicker1.jmdat
User data location: /home/waxwing/.joinmarket/
2021-02-17 16:05:10,742 [DEBUG]  rpc: getblockchaininfo []
Enter passphrase to decrypt wallet: 
2021-02-17 16:05:13,416 [DEBUG]  requesting detailed wallet history
2021-02-17 16:05:13,465 [DEBUG]  rpc: listlabels []
2021-02-17 16:05:13,466 [DEBUG]  rpc: getaddressesbylabel ['joinmarket-wallet-9d5804']
2021-02-17 16:05:13,472 [DEBUG]  got used indices: {0: [3, 12], 1: [0, 2], 2: [0, 0], 3: [0, 0], 4: [0, 0]}
2021-02-17 16:05:13,514 [DEBUG]  rpc: listlabels []
2021-02-17 16:05:13,515 [DEBUG]  rpc: getaddressesbylabel ['joinmarket-wallet-9d5804']
2021-02-17 16:05:13,515 [DEBUG]  Wallet successfully synced
2021-02-17 16:05:13,516 [DEBUG]  rpc: listunspent [0]
2021-02-17 16:05:13,518 [DEBUG]  bitcoind sync_unspent took 0.0024628639221191406sec
Found a snicker tx: 0afea91a2f9b0ae4e67ef45ea5f1e3e72a7866070de6394cc5c9343eb6518af8
Found a snicker tx: a54fddd20f2cb2cce051b8e7d6138cdd8154ecb5189e6a692c6d7d1dae2bc301
Found a snicker tx: 13ba85c886d09eaddea75c9cc706dfb9084759f93903423087c31b86056effdc
Found a snicker tx: 762a697e647ca7642317c860508dfd92829a59a180b35cc0c7047c77ad696c15
Script finished, recovery is complete.
```

and after I was able to successfully find the same balance, and the same two imported keys containing coins exactly as the pre-recovery wallet, using `wallet-tool`.

This process is all quite laborious and messy, but that's exactly what's expected as commented above. This case should be very rare. I'll add a comment to the docs to ensure people understand that this is to be avoided with wallet backups.

AdamISZ commented:
Rebased on master and squashed. I plan to merge this soon, I think I am done testing.

I would really appreciate even a small amount of code review (for example, you could just check the changes to the jmbitcoin package, or just look at the new tests).

Even though nothing here will be enabled after the merge (default setting is false), and so users not interested will see zero changes, it'll be really helpful so I can finally address #748 (and also, though pie in the sky, could open us to using a more peer to peer communication model, elsewhere).

AdamISZ commented:
Added one final commit - disables the most problematic (given lack of testing) function, namely passive receipt of SNICKERs withing a yield generator on mainnet. And a reminder that of course the entire functionality is switched off by default, also, and mainnet is disrecommended in the docs, in any use pattern.

I consider this complete and tested well enough for people to use in test scenarios, and given all the above, merging.
